flytekit/types/file/file.py

```diff
@@ -324,6 +325,44 @@ def assert_type(
     def get_literal_type(self, t: typing.Union[typing.Type[FlyteFile], os.PathLike]) -> LiteralType:
         return LiteralType(blob=self._blob_type(format=FlyteFilePathTransformer.get_format(t)))
 
+    def get_mime_type_from_python_type(self, extension: str) -> str:
+        extension_to_mime_type = {
+            "html": mimetypes.types_map[".html"],
+            "jpeg": mimetypes.types_map[".jpeg"],
+            "png": mimetypes.types_map[".png"],
+            "pdf": mimetypes.types_map[".pdf"],
+            "txt": mimetypes.types_map[".txt"],
+            "csv": mimetypes.types_map[".csv"],
+            "svg": mimetypes.types_map[".svg"],
+            "hdf5": "text/plain",
+            "joblib": "application/octet-stream",
+            "python_pickle": "application/octet-stream",
+            "ipynb": "application/json",
+            "onnx": "application/json",
+            "tfrecord": "application/octet-stream",
+        }
+        return extension_to_mime_type[extension]
+
+    def validate_file_type(self, python_type: typing.Type[FlyteFile], source_path: typing.Union[str, os.PathLike]):
+        try:
+            # isolate the exception to the libmagic import
+            import magic
+
+        except ImportError as e:
+            logger.warning(f"Libmagic is not installed. Error message: {e}")
+            return
+
+        if FlyteFilePathTransformer.get_format(python_type):
+            try:
+                real_type = magic.from_file(source_path, mime=True)
+                expected_type = self.get_mime_type_from_python_type(FlyteFilePathTransformer.get_format(python_type))
+                if real_type != expected_type:
+                    raise ValueError(f"Incorrect file type, expected {expected_type}, got {real_type}")
+            except FileNotFoundError:
+                # This scenario occurs during unit testing.
+                # If the actual file does not exist, no comparison of file types is needed.
```

Comment:
what test caused this?

Reply:
Because in the [test_flyte_file_in_dataclassjsonmixin unit test](https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/tests/flytekit/unit/core/test_type_engine.py#L1045-L1074), the [validate_file_type](https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/flytekit/types/file/file.py#L346-L375) method will be passed, in which [magic.from_file](https://github.com/flyteorg/flytekit/blob/b176572a0c97cefd64bbe0c28a6401839be3f134/flytekit/types/file/file.py#L368) will check the file. If the file does not exist, a **FileNotFoundException** will be thrown.

Reply:
as shown in the picture
![Screenshot 2023-10-31 at 2 51 22 PM](https://github.com/flyteorg/flytekit/assets/14800485/074453eb-76df-4994-8326-421c81513d9c)

Reply:
That’s why I used **`try-except`** to handle it. Is it appropriate to handle it this way? Is there any other better way?

Reply:
After thinking for a while, I think it is not right to write try-except in order to let the unit test pass. Maybe we should adjust the unit test.
I think the simple thing to do is to skip this test if python-magic is installed, what do you think?

Reply:
I removed the try-except and adjusted the unit tests.
https://github.com/flyteorg/flytekit/blob/7ac4c8f840fb587d9a1df3ed0aaaf40553a8e485/tests/flytekit/unit/core/test_type_engine.py#L1047-L1080

Reply:
One of the usecases for FlyteFile is to point to remote files, in other words, you might have access to a remote file (e.g. in s3) that you want to pass to a Flyte workflow. 

More concretely, we should only validate FlyteFiles for which their `path` is considered local (via a call to [FileAccessProvider.is_remote](https://github.com/flyteorg/flytekit/blob/master/flytekit/core/data_persistence.py#L184-L191)).

Reply:
OK, I updated it.
like this
https://github.com/flyteorg/flytekit/blob/c1e2774cca85f84f9d7a0a80d140c5591ae0bd54/tests/flytekit/unit/core/test_type_engine.py#L1010-L1019

Reply:
https://github.com/flyteorg/flytekit/blob/06bd4c4d2cdfdd46119342c58e51e69cc8687360/flytekit/types/file/file.py#L358-L359